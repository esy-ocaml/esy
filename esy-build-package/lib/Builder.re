let relocateSourcePath = (config: Config.t, spec: BuildSpec.t) => {
  let cmd =
    Bos.Cmd.(
      empty
      % config.rsyncCmd
      % "--quiet"
      % "--archive"
      % "--exclude"
      % p(spec.buildPath)
      % "--exclude"
      % "node_modules"
      % "--exclude"
      % "_build"
      % "--exclude"
      % "_release"
      % "--exclude"
      % "_esybuild"
      % "--exclude"
      % "_esyinstall"
      /* The trailing "/" is important as it makes rsync to sync the contents of
       * origPath rather than the origPath itself into destPath, see "man rsync" for
       * details.
       */
      % (Path.to_string(spec.sourcePath) ++ "/")
      % p(spec.buildPath)
    );
  Bos.OS.Cmd.run(cmd);
};

let commitBuildToStore = (config: Config.t, spec: BuildSpec.t) => {
  open Run;
  let rewritePrefixInFile = (~origPrefix, ~destPrefix, path) => {
    let cmd =
      Bos.Cmd.(
        empty
        % config.fastreplacestringCmd
        % p(path)
        % p(origPrefix)
        % p(destPrefix)
      );
    Bos.OS.Cmd.run(cmd);
  };
  let rewriteTargetInSymlink = (~origPrefix, ~destPrefix, path) => {
    let%bind targetPath = symlink_target(path);
    switch (Path.rem_prefix(origPrefix, targetPath)) {
    | Some(basePath) =>
      let nextTargetPath = Path.append(destPrefix, basePath);
      let%bind () = rm(path);
      let%bind () = symlink(~target=nextTargetPath, path);
      ok;
    | None => ok
    };
  };
  let relocate = (path: Path.t, stats: Unix.stats) =>
    switch stats.st_kind {
    | Unix.S_REG =>
      rewritePrefixInFile(
        ~origPrefix=spec.stagePath,
        ~destPrefix=spec.installPath,
        path
      )
    | Unix.S_LNK =>
      rewriteTargetInSymlink(
        ~origPrefix=spec.stagePath,
        ~destPrefix=spec.installPath,
        path
      )
    | _ => Ok()
    };
  let%bind () =
    Bos.OS.File.write(
      Path.(spec.stagePath / "_esy" / "storePrefix"),
      Path.to_string(config.storePath)
    );
  let%bind () = traverse(spec.stagePath, relocate);
  let%bind () = Bos.OS.Path.move(spec.stagePath, spec.installPath);
  ok;
};

let relocateBuildPath = (_config: Config.t, spec: BuildSpec.t) => {
  open Run;
  let savedBuild = spec.buildPath / "_build";
  let currentBuild = spec.sourcePath / "_build";
  let backupBuild = spec.sourcePath / "_build.prev";
  let start = (_config, _spec) => {
    let%bind () =
      if%bind (exists(currentBuild)) {
        mv(currentBuild, backupBuild);
      } else {
        ok;
      };
    let%bind () = mkdir(savedBuild);
    let%bind () = mv(savedBuild, currentBuild);
    ok;
  };
  let commit = (_config, _spec) => {
    let%bind () =
      if%bind (exists(currentBuild)) {
        mv(currentBuild, savedBuild);
      } else {
        ok;
      };
    let%bind () =
      if%bind (exists(backupBuild)) {
        mv(backupBuild, currentBuild);
      } else {
        ok;
      };
    ok;
  };
  (start, commit);
};

let findSourceModTime = (spec: BuildSpec.t) => {
  open Run;
  let visit = (path: Path.t) =>
    fun
    | Ok(maxTime) =>
      if (path == spec.sourcePath) {
        Ok(maxTime);
      } else {
        let%bind {Unix.st_mtime: time, _} = Bos.OS.Path.symlink_stat(path);
        Ok(time > maxTime ? time : maxTime);
      }
    | error => error;
  let traverse =
    `Sat(
      path =>
        switch (Path.basename(path)) {
        | "node_modules" => Ok(false)
        | "_esy" => Ok(false)
        | "_release" => Ok(false)
        | "_build" => Ok(false)
        | "_install" => Ok(false)
        | base when base.[0] == '.' => Ok(false)
        | _ => Ok(true)
        }
    );
  Result.join(
    Bos.OS.Path.fold(
      ~dotfiles=true,
      ~traverse,
      visit,
      Ok(0.),
      [spec.sourcePath]
    )
  );
};

let doNothing = (_config: Config.t, _spec: BuildSpec.t) => Run.ok;

/**
 * Execute `run` within the build environment for `spec`.
 */
let withBuildEnv = (~commit=false, config: Config.t, spec: BuildSpec.t, f) => {
  open Run;
  let {BuildSpec.sourcePath, installPath, buildPath, stagePath, _} = spec;
  let (rootPath, prepareRootPath, completeRootPath) =
    switch (spec.buildType, spec.sourceType) {
    | (InSource, _) => (buildPath, relocateSourcePath, doNothing)
    | (JbuilderLike, Immutable) => (buildPath, relocateSourcePath, doNothing)
    | (JbuilderLike, Transient) =>
      let (start, commit) = relocateBuildPath(config, spec);
      (sourcePath, start, commit);
    | (JbuilderLike, Root) => (sourcePath, doNothing, doNothing)
    | (OutOfSource, _) => (sourcePath, doNothing, doNothing)
    };
  let%bind sandboxConfig = {
    open Sandbox;
    let regex = (base, segments) => {
      let pat =
        String.concat(Path.dir_sep, [Path.to_string(base), ...segments]);
      Regex(pat);
    };
    let%bind tempPath = {
      let v = Path.v(Bos.OS.Env.opt_var("TMPDIR", ~absent="/tmp"));
      let%bind v = realpath(v);
      Ok(Path.to_string(v));
    };
    let allowWriteToSourcePath =
      switch spec.buildType {
      | JbuilderLike => [
          Subpath(Path.to_string(sourcePath / "_build")),
          regex(sourcePath, [".*", "[^/]*\\.install"]),
          regex(sourcePath, ["[^/]*\\.install"]),
          regex(sourcePath, [".*", "[^/]*\\.opam"]),
          regex(sourcePath, ["[^/]*\\.opam"]),
          regex(sourcePath, [".*", "jbuild-ignore"])
        ]
      | _ => []
      };
    Ok(
      allowWriteToSourcePath
      @ [
        regex(sourcePath, [".*", "\\.merlin"]),
        regex(sourcePath, ["\\.merlin"]),
        Subpath(Path.to_string(buildPath)),
        Subpath(Path.to_string(stagePath)),
        Subpath("/private/tmp"),
        Subpath("/tmp"),
        Subpath(tempPath)
      ]
    );
  };
  let env =
    switch (Bos.OS.Env.var("TERM")) {
    | Some(term) => Astring.String.Map.add("TERM", term, spec.env)
    | None => spec.env
    };
  let%bind exec = Sandbox.sandboxExec({allowWrite: sandboxConfig});
  let path =
    switch (Astring.String.Map.find("PATH", env)) {
    | Some(path) => String.split_on_char(':', path)
    | None => []
    };
  let run = cmd => {
    let%bind cmd = Cmd.resolveInvocation(path, cmd);
    Bos.OS.Cmd.(
      in_null |> exec(~err=Bos.OS.Cmd.err_run_out, ~env, cmd) |> to_stdout
    );
  };
  let runInteractive = cmd => {
    let%bind cmd = Cmd.resolveInvocation(path, cmd);
    Bos.OS.Cmd.(
      in_stdin |> exec(~err=Bos.OS.Cmd.err_stderr, ~env, cmd) |> to_stdout
    );
  };
  /*
   * Prepare build/install.
   */
  let prepare = () => {
    let%bind () = rmdir(installPath);
    let%bind () = rmdir(stagePath);
    let%bind () = mkdir(stagePath);
    let%bind () = mkdir(stagePath / "bin");
    let%bind () = mkdir(stagePath / "lib");
    let%bind () = mkdir(stagePath / "etc");
    let%bind () = mkdir(stagePath / "sbin");
    let%bind () = mkdir(stagePath / "man");
    let%bind () = mkdir(stagePath / "share");
    let%bind () = mkdir(stagePath / "doc");
    let%bind () = mkdir(stagePath / "_esy");
    let%bind () =
      switch (spec.sourceType, spec.buildType) {
      | (Immutable, _)
      | (_, InSource) =>
        let%bind () = rmdir(buildPath);
        let%bind () = mkdir(buildPath);
        ok;
      | _ =>
        let%bind () = mkdir(buildPath);
        ok;
      };
    let%bind () = prepareRootPath(config, spec);
    let%bind () = mkdir(buildPath / "_esy");
    ok;
  };
  /*
   * Finalize build/install.
   */
  let finalize = result =>
    switch result {
    | Ok () =>
      let%bind () =
        if (commit) {
          commitBuildToStore(config, spec);
        } else {
          ok;
        };
      let%bind () = completeRootPath(config, spec);
      ok;
    | error =>
      let%bind () = completeRootPath(config, spec);
      error;
    };
  let%bind () = Store.init(config.storePath);
  let%bind () = Store.init(config.localStorePath);
  let%bind () = prepare();
  let result = withCwd(rootPath, ~f=f(run, runInteractive));
  let%bind () = finalize(result);
  result;
};

let build =
    (~buildOnly=true, ~force=false, config: Config.t, spec: BuildSpec.t) => {
  open Run;
  Logs.debug(m => m("start %s", spec.id));
  let performBuild = sourceModTime => {
    Logs.debug(m => m("building"));
    let runBuildAndInstall = (run, _runInteractive, ()) => {
      let runList = cmds => {
        let rec _runList = cmds =>
          switch cmds {
          | [] => Ok()
          | [cmd, ...cmds] =>
            Logs.app(m =>
              m("# esy-build-package: running: %s", Bos.Cmd.to_string(cmd))
            );
            switch (run(cmd)) {
            | Ok(_) => _runList(cmds)
            | Error(err) => Error(err)
            };
          };
        _runList(cmds);
      };
      let {BuildSpec.build, install, _} = spec;
      let%bind () = runList(build);
      let%bind () =
        if (! buildOnly) {
          runList(install);
        } else {
          ok;
        };
      ok;
    };
    let startTime = Unix.gettimeofday();
    let%bind () =
      withBuildEnv(~commit=! buildOnly, config, spec, runBuildAndInstall);
    let%bind info = {
      let%bind sourceModTime =
        switch (sourceModTime, spec.sourceType) {
        | (None, BuildSpec.Root)
        | (None, BuildSpec.Transient) =>
          let%bind v = findSourceModTime(spec);
          Ok(Some(v));
        | (v, _) => Ok(v)
        };
      Ok(
        BuildInfo.{sourceModTime, timeSpent: Unix.gettimeofday() -. startTime}
      );
    };
    BuildInfo.write(spec, info);
  };
  switch (force, spec.sourceType) {
  | (true, _) =>
    Logs.debug(m => m("forcing build"));
    performBuild(None);
  | (false, BuildSpec.Transient)
  | (false, BuildSpec.Root) =>
    Logs.debug(m => m("checking for staleness"));
    let info = BuildInfo.read(spec);
    let prevSourceModTime =
      Option.bind(~f=v => v.BuildInfo.sourceModTime, info);
    let%bind sourceModTime = findSourceModTime(spec);
    switch prevSourceModTime {
    | Some(prevSourceModTime) when sourceModTime > prevSourceModTime =>
      performBuild(Some(sourceModTime))
    | None => performBuild(Some(sourceModTime))
    | Some(_) =>
      Logs.debug(m => m("source code is not modified, skipping"));
      ok;
    };
  | (false, BuildSpec.Immutable) =>
    let%bind installPathExists = exists(spec.installPath);
    if (installPathExists) {
      Logs.debug(m => m("build exists in store, skipping"));
      ok;
    } else {
      performBuild(None);
    };
  };
};
