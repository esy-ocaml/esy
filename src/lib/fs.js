/**
 * @flow
 */

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');
const os = require('os');

import walkDir from 'walkdir';
import {copy, copySync} from 'fs-extra';

import {promisify} from './Promise';

// Promote some of the fs built-ins
export const stat: (path: string) => Promise<fs.Stats> = promisify(fs.stat);
export const lstat: (path: string) => Promise<fs.Stats> = promisify(fs.lstat);
export const readdir: (path: string, opts: void) => Promise<Array<string>> = promisify(
  fs.readdir,
);
export const rename: (oldPath: string, newPath: string) => Promise<void> = promisify(
  fs.rename,
);
export const exists: (path: string) => Promise<boolean> = promisify(fs.exists, true);
export const realpath: (p: string) => Promise<string> = promisify(fs.realpath);
export const readlink: (p: string) => Promise<string> = promisify(fs.readlink);
export const realpathSync: (p: string) => string = fs.realpathSync;
export const unlink: (prefix: string) => Promise<string> = promisify(fs.unlink);
export const readFileBuffer: (p: string) => Promise<Buffer> = promisify(fs.readFile);
export const writeFile: (
  path: string,
  data: Buffer | string,
  encdogin?: string,
) => Promise<void> = promisify(fs.writeFile);
export const chmod: (path: string, mode: number | string) => Promise<void> = promisify(
  fs.chmod,
);
export const writeFileSync = fs.writeFileSync;

// Promote 3rd-party fs utils
export const rmdir: (p: string) => Promise<void> = promisify(require('rimraf'));
export const rmdirSync: (p: string) => void = require('rimraf').sync;
export const mkdirp: (path: string) => Promise<void> = promisify(require('mkdirp'));

// mkdtemp
export const _mkdtemp: string => Promise<string> = promisify(fs.mkdtemp);
export const _mkdtempSync = fs.mkdtempSync;

export function mkdtemp(prefix: string) {
  const root = os.tmpdir();
  return _mkdtemp(path.join(root, prefix));
}

// copydir
const _copydir: (
  string,
  string,
  {filter?: string => boolean},
) => Promise<void> = promisify(copy);

export {copy};

export async function copydir(
  from: string,
  to: string,
  params?: {
    exclude?: string[],
  } = {},
): Promise<void> {
  await _copydir(from, to, {
    filter: filename => !(params.exclude && params.exclude.includes(filename)),
  });
}

export function copydirSync(
  from: string,
  to: string,
  params?: {
    exclude?: string[],
  } = {},
): void {
  copySync(from, to, {
    filter: filename => !(params.exclude && params.exclude.includes(filename)),
  });
}

export async function readFile(p: string): Promise<string> {
  const data = await readFileBuffer(p);
  return data.toString('utf8');
}

export function readFileSync(p: string) {
  return fs.readFileSync(p, 'utf8');
}

export async function readJson(p: string, parse: string => Function = JSON.parse) {
  const data = await readFile(p);
  return parse(data);
}

export function readJsonSync(p: string, parse: string => Function = JSON.parse) {
  const data = readFileSync(p);
  return parse(data);
}

export const fsSymlink: (
  target: string,
  path: string,
  type?: 'dir' | 'file' | 'junction',
) => Promise<void> = promisify(fs.symlink);

export async function symlink(src: string, dest: string): Promise<void> {
  try {
    const stats = await lstat(dest);

    if (stats.isSymbolicLink() && (await exists(dest))) {
      const resolved = await realpath(dest);
      if (resolved === src) {
        return;
      }
    }

    await unlink(dest);
  } catch (err) {
    if (err.code !== 'ENOENT') {
      throw err;
    }
  }

  try {
    if (process.platform === 'win32') {
      // use directory junctions if possible on win32, this requires absolute paths
      await fsSymlink(src, dest, 'junction');
    } else {
      // use relative paths otherwise which will be retained if the directory is moved
      const relative = path.relative(
        fs.realpathSync(path.dirname(dest)),
        fs.realpathSync(src),
      );
      await fsSymlink(relative, dest);
    }
  } catch (err) {
    if (err.code === 'EEXIST') {
      // race condition
      await symlink(src, dest);
    } else {
      throw err;
    }
  }
}

export type WalkFiles = Array<{
  relative: string,
  absolute: string,
  basename: string,
  mtime: number,
  stats: fs.Stats,
}>;

export async function walk(
  dir: string,
  relativeDir?: ?string,
  ignoreBasenames?: Set<string> = new Set(),
): Promise<WalkFiles> {
  let files = [];

  let filenames = await readdir(dir);
  if (ignoreBasenames.size) {
    filenames = filenames.filter(name => !ignoreBasenames.has(name));
  }

  for (const name of filenames) {
    const relative = relativeDir ? path.join(relativeDir, name) : name;
    const loc = path.join(dir, name);
    const stats = await lstat(loc);

    files.push({
      relative,
      basename: name,
      absolute: loc,
      mtime: +stats.mtime,
      stats,
    });

    if (stats.isDirectory()) {
      files = files.concat(await walk(loc, relative, ignoreBasenames));
    }
  }

  return files;
}

export function findMaxMtime(
  dirname: string,
  options?: {ignore?: string => boolean} = {},
): Promise<number> {
  const ignore = options.ignore ? options.ignore : _filename => false;
  let maxMtime = -Infinity;

  return new Promise((resolve, reject) => {
    const w = walkDir(dirname);
    w.on('path', (name, stat) => {
      if (ignore(name)) {
        w.ignore(name);
      } else if (stat.isFile()) {
        const mtime = Math.floor(stat.mtime.getTime() / 1000);
        if (mtime > maxMtime) {
          maxMtime = mtime;
        }
      }
    });
    w.on('end', () => {
      resolve(maxMtime);
    });
    w.on('error', error => {
      reject(error);
    });
  });
}
