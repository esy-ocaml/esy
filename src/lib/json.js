/**
 * @flow
 */

const jsonStableStringify = require('json-stable-stringify');

export function stableStringifyPretty(obj: any): string {
  return jsonStableStringify(obj, {space: '  '});
}

export const parse = JSON.parse;
export const stringify = JSON.stringify;
