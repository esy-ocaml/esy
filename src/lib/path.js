/**
 * @flow
 */

import invariant from 'invariant';
const path = require('path');

/**
 * Absolute path.
 */
export opaque type AbsolutePath: ConcretePath = string;

/**
 * Path with no `.` or `..` or any bash variables in it.
 */
export opaque type ConcretePath: Path = string;

/**
 * Path which can contain bash/make variables in it, cannot be used in
 * filesystem operations.
 */
export opaque type AbstractPath: Path = string;

export opaque type Path: string = AbsolutePath | ConcretePath | AbstractPath;

export function absolute(p: string): AbsolutePath {
  p = sanitizeAbsolutePath(p);
  return p;
}

export function concrete(p: string): ConcretePath {
  p = sanitizeConcretePath(p);
  return p;
}

export function abstract(p: string): AbstractPath {
  return p;
}

export function join<B: Path | string>(base: B, ...p: string[]): B {
  return (path.join(base, ...p): any);
}

export function basename<P: Path | string>(p: P): string {
  return path.basename(p);
}

export function dirname<P: Path | string>(p: P): P {
  return (path.dirname(p): any);
}

export function length(p: ConcretePath): number {
  return p.length;
}

export function toString(p: Path | string): string {
  return p;
}

export const sep: typeof path.sep = path.sep;

export const relative: typeof path.relative = path.relative;

function sanitizeAbsolutePath(p) {
  p = sanitizeConcretePath(p);
  invariant(path.isAbsolute(p), 'Should be absolute path but got: %s', p);
  return p;
}

function sanitizeConcretePath(p) {
  // NOTE: regexpes are stateful, this is why we are creating them here.
  // Consider resetting their state after matching instead.
  const REMOVE_TRAILING_SLASH_RE = /\/+$/g;
  const ENVIRONMENT_VAR_RE = /\$[a-zA-Z_]/g;
  const DOT_PATH_SEGMENT_RE = /\/\.\//g;
  const DOT_DOT_PATH_SEGMENT_RE = /\/\.\.\//g;
  const SANITIZE_SLASH_RE = /\/+/g;

  invariant(
    DOT_PATH_SEGMENT_RE.exec(p) == null,
    'Should not contain "." segments: %s',
    p,
  );
  invariant(
    DOT_DOT_PATH_SEGMENT_RE.exec(p) == null,
    'Should not contain ".." segments: %s',
    p,
  );
  invariant(
    ENVIRONMENT_VAR_RE.exec(p) == null,
    'Path should not contain variable references: %s',
    p,
  );
  p = path.normalize(p);
  p = p.replace(REMOVE_TRAILING_SLASH_RE, '');
  p = p === '' ? '/' : p;
  return p;
}
