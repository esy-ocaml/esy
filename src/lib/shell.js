/**
 * @flow
 */

import shellEscape from 'shell-escape';

export function singleQuote(v: string): string {
  return shellEscape([v]);
}

export function doubleQuote(v: string): string {
  return JSON.stringify(v);
}
