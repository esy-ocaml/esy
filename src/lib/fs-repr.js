/**
 * @flow
 */

import invariant from 'invariant';
import * as path from 'path';
import * as fs from './fs';

export type FileNode = {
  type: 'file',
  name: string,
  content: ?any,
};

export type LinkNode = {
  type: 'link',
  name: string,
  realpath: string,
};

export type DirectoryNode = {
  type: 'directory',
  name: string,
  nodes: Node[],
};

export type Node = FileNode | LinkNode | DirectoryNode;

export function file(name: string, content: ?any): FileNode {
  return {type: 'file', name, content};
}

export function link(name: string, realpath: string): LinkNode {
  return {type: 'link', name, realpath};
}

export function directory(name: string, nodes: Node[]): DirectoryNode {
  return {type: 'directory', name, nodes};
}

export async function write(rootDirname: string, nodes: Node[]): Promise<void> {
  invariant((await fs.readdir(rootDirname)).length === 0, 'Directory is not empty');

  async function writeFile(pathname: string, node: FileNode) {
    const content =
      typeof node.content === 'string' || node.content instanceof Buffer
        ? node.content
        : JSON.stringify(node.content);
    await fs.writeFile(pathname, content);
  }

  async function writeLink(pathname: string, node: LinkNode) {
    await fs.symlink(path.resolve(node.realpath, rootDirname), pathname);
  }

  async function writeDirectory(pathname: string, node: DirectoryNode) {
    await fs.mkdirp(pathname);
    const tasks = [];
    for (const nextNode of node.nodes) {
      const nextPathname = path.join(pathname, nextNode.name);
      if (nextNode.type === 'file') {
        tasks.push(writeFile(nextPathname, nextNode));
      } else if (nextNode.type === 'link') {
        tasks.push(writeLink(nextPathname, nextNode));
      } else if (nextNode.type === 'directory') {
        tasks.push(writeDirectory(nextPathname, nextNode));
      }
    }
    await Promise.all(tasks);
  }

  await writeDirectory(rootDirname, {type: 'directory', name: '<root>', nodes});
}

export type ReadOptions = {
  filter?: (filename: string) => boolean,
};

export async function read(
  rootDirname: string,
  options?: ReadOptions = {},
): Promise<Node[]> {
  const {filter} = options;

  function crawlFile(pathname: string, name: string): Promise<FileNode> {
    return Promise.resolve({
      type: 'file',
      name,
      content: null,
    });
  }

  async function crawlLink(pathname: string, name: string): Promise<LinkNode> {
    const realpath = path.relative(await fs.realpath(pathname), rootDirname);
    return {
      type: 'link',
      name,
      realpath,
    };
  }

  async function crawlDirectory(
    pathname: string,
    name: string,
    relpath: string,
  ): Promise<DirectoryNode> {
    const tasks: Array<Promise<Node>> = [];
    for (const n of await fs.readdir(pathname)) {
      const nextPathname = path.join(pathname, n);
      const nextRelpath = path.join(relpath, n);
      if (filter != null && !filter(nextRelpath)) {
        continue;
      }
      const stat = await fs.stat(nextPathname);
      if (stat.isSymbolicLink()) {
        tasks.push(crawlLink(nextPathname, n));
      } else if (stat.isFile()) {
        tasks.push(crawlFile(nextPathname, n));
      } else if (stat.isDirectory()) {
        tasks.push(crawlDirectory(nextPathname, n, nextRelpath));
      }
    }
    const nodes = await Promise.all(tasks);
    nodes.sort((a, b) => a.name.localeCompare(b.name));
    return {type: 'directory', name, nodes};
  }

  const {nodes} = await crawlDirectory(rootDirname, '<root>', '<root>');
  return nodes;
}
