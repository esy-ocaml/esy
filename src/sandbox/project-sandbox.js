/**
 * @flow
 */

import type {BuildSpec, Sandbox} from '../types';

import * as path from 'path';

import * as M from '../package-manifest';
import * as Crawl from './crawl';

export type Options = {
  forRelease?: boolean,
};

/**
 * Create a project sandbox.
 */
export async function create(
  sandboxPath: string,
  options: Options = {},
): Promise<Sandbox> {
  const manifestResolutionCache = new Map();

  function resolveManifestCached(dep, context): Promise<?Crawl.Resolution> {
    const baseDirectory = context.sourcePath;
    const key = `${baseDirectory}__${dep.name}`;
    let resolution = manifestResolutionCache.get(key);
    if (resolution == null) {
      resolution = M.resolve(dep.name, baseDirectory).then(res => {
        if (res == null) {
          return res;
        }
        return {
          manifest: res.manifest,
          sourcePath: path.dirname(res.filename),
        };
      });
      manifestResolutionCache.set(key, resolution);
    }
    return resolution;
  }

  const buildCache: Map<string, Promise<BuildSpec>> = new Map();

  function crawlBuildCached(context): Promise<BuildSpec> {
    const key = context.sourcePath;
    let build = buildCache.get(key);
    if (build == null) {
      build = Crawl.crawlBuild(context);
      buildCache.set(key, build);
    }
    return build;
  }

  const env = Crawl.getDefaultEnvironment();

  const {manifest, filename: manifestFilename} = await M.read(sandboxPath);

  const crawlContext: Crawl.SandboxCrawlContext = {
    manifest,
    env,
    sourcePath: sandboxPath,
    sandboxPath,
    resolveManifest: resolveManifestCached,
    crawlBuild: crawlBuildCached,
    dependencyTrace: [],
    options,
  };

  const root = await crawlBuildCached(crawlContext);

  const devDependenciesReqs = Crawl.dependenciesFromObj('dev', manifest.devDependencies);
  const {dependencies: devDependencies} = await Crawl.crawlDependencies(
    devDependenciesReqs,
    crawlContext,
  );

  return {env, devDependencies, root};
}
