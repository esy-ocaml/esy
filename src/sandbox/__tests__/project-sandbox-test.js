/**
 * @flow
 */

import * as fs from '../../lib/fs';
import * as FSRepr from '../../lib/fs-repr';
import {create} from '../project-sandbox';

function pkg(packageJson, ...dependencies) {
  const nodes = [FSRepr.file('package.json', packageJson)];
  if (dependencies.length > 0) {
    nodes.push(FSRepr.directory('node_modules', dependencies));
  }
  return FSRepr.directory(packageJson.name, nodes);
}

describe('build-sandbox', function() {
  let directoriesToCleanup = [];

  async function prepareSandbox(nodes: FSRepr.Node[]): Promise<string> {
    const tempdir = await fs.mkdtemp('esy-test-sandbox');
    directoriesToCleanup.push(tempdir);
    await FSRepr.write(tempdir, nodes);
    return tempdir;
  }

  beforeEach(function() {
    directoriesToCleanup = [];
  });

  afterEach(async function() {
    for (const dirname of directoriesToCleanup) {
      try {
        await fs.unlink(dirname);
      } catch (_err) {}
    }
  });

  test('builds a sandbox from a single package', async function() {
    const sandboxDir = await prepareSandbox(
      pkg({
        name: 'app',
        version: '0.1.0',
        _resolved: 'app',
      }).nodes,
    );
    const sandbox = await create(sandboxDir);
    expect(sandbox.root).toMatchSnapshot();
  });

  test('builds a sandbox from a package with deps', async function() {
    const sandboxDir = await prepareSandbox(
      pkg(
        {
          name: 'app',
          version: '0.1.0',
          dependencies: {
            dep: '*',
          },
          _resolved: 'app',
        },
        pkg({
          name: 'dep',
          version: '0.1.0',
          _resolved: 'dep',
        }),
      ).nodes,
    );
    const sandbox = await create(sandboxDir);
    expect(sandbox.root).toMatchSnapshot();
  });

  test('error: missing a dep', async function() {
    const sandboxDir = await prepareSandbox(
      pkg({
        name: 'app',
        version: '0.1.0',
        dependencies: {
          dep: '*',
        },
        _resolved: 'app',
      }).nodes,
    );
    const sandbox = await create(sandboxDir);
    expect(sandbox.root).toMatchSnapshot();
  });

  test('error: circular deps', async function() {
    const sandboxDir = await prepareSandbox(
      pkg(
        {
          name: 'app',
          version: '0.1.0',
          dependencies: {
            dep: '*',
          },
          _resolved: 'app',
        },
        pkg({
          name: 'dep',
          version: '0.1.0',
          dependencies: {
            app: '*',
          },
          _resolved: 'dep',
        }),
      ).nodes,
    );
    const sandbox = await create(sandboxDir);
    expect(sandbox.root).toMatchSnapshot();
  });
});
