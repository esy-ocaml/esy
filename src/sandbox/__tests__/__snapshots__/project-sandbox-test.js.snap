// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build-sandbox builds a sandbox from a package with deps 1`] = `
Object {
  "buildCommand": Array [],
  "buildType": "out-of-source",
  "dependencies": Map {
    "dep-0.1.0" => Object {
      "buildCommand": Array [],
      "buildType": "out-of-source",
      "dependencies": Map {},
      "errors": Array [],
      "exportedEnv": Object {},
      "id": "dep-0.1.0",
      "installCommand": Array [],
      "name": "dep",
      "packageJson": Object {
        "dependencies": Object {},
        "devDependencies": Object {},
        "esy": Object {
          "build": null,
          "buildsInSource": false,
          "exportedEnv": Object {},
          "install": null,
        },
        "name": "dep",
        "peerDependencies": Object {},
        "version": "0.1.0",
      },
      "shouldBePersisted": false,
      "sourcePath": "node_modules/dep",
      "sourceType": "transient",
      "version": "0.1.0",
    },
  },
  "errors": Array [],
  "exportedEnv": Object {},
  "id": "app-0.1.0",
  "installCommand": Array [],
  "name": "app",
  "packageJson": Object {
    "dependencies": Object {
      "dep": "*",
    },
    "devDependencies": Object {},
    "esy": Object {
      "build": null,
      "buildsInSource": false,
      "exportedEnv": Object {},
      "install": null,
    },
    "name": "app",
    "peerDependencies": Object {},
    "version": "0.1.0",
  },
  "shouldBePersisted": false,
  "sourcePath": "",
  "sourceType": "root",
  "version": "0.1.0",
}
`;

exports[`build-sandbox builds a sandbox from a single package 1`] = `
Object {
  "buildCommand": Array [],
  "buildType": "out-of-source",
  "dependencies": Map {},
  "errors": Array [],
  "exportedEnv": Object {},
  "id": "app-0.1.0",
  "installCommand": Array [],
  "name": "app",
  "packageJson": Object {
    "dependencies": Object {},
    "devDependencies": Object {},
    "esy": Object {
      "build": null,
      "buildsInSource": false,
      "exportedEnv": Object {},
      "install": null,
    },
    "name": "app",
    "peerDependencies": Object {},
    "version": "0.1.0",
  },
  "shouldBePersisted": false,
  "sourcePath": "",
  "sourceType": "root",
  "version": "0.1.0",
}
`;

exports[`build-sandbox error: circular deps 1`] = `
Object {
  "buildCommand": Array [],
  "buildType": "out-of-source",
  "dependencies": Map {
    "dep-0.1.0" => Object {
      "buildCommand": Array [],
      "buildType": "out-of-source",
      "dependencies": Map {},
      "errors": Array [
        Object {
          "message": "Circular dependency \\"app\\" found
  At app -> dep",
        },
      ],
      "exportedEnv": Object {},
      "id": "dep-0.1.0",
      "installCommand": Array [],
      "name": "dep",
      "packageJson": Object {
        "dependencies": Object {
          "app": "*",
        },
        "devDependencies": Object {},
        "esy": Object {
          "build": null,
          "buildsInSource": false,
          "exportedEnv": Object {},
          "install": null,
        },
        "name": "dep",
        "peerDependencies": Object {},
        "version": "0.1.0",
      },
      "shouldBePersisted": false,
      "sourcePath": "node_modules/dep",
      "sourceType": "transient",
      "version": "0.1.0",
    },
  },
  "errors": Array [
    Object {
      "message": "Circular dependency \\"app\\" found
  At app -> dep",
    },
  ],
  "exportedEnv": Object {},
  "id": "app-0.1.0",
  "installCommand": Array [],
  "name": "app",
  "packageJson": Object {
    "dependencies": Object {
      "dep": "*",
    },
    "devDependencies": Object {},
    "esy": Object {
      "build": null,
      "buildsInSource": false,
      "exportedEnv": Object {},
      "install": null,
    },
    "name": "app",
    "peerDependencies": Object {},
    "version": "0.1.0",
  },
  "shouldBePersisted": false,
  "sourcePath": "",
  "sourceType": "root",
  "version": "0.1.0",
}
`;

exports[`build-sandbox error: missing a dep 1`] = `
Object {
  "buildCommand": Array [],
  "buildType": "out-of-source",
  "dependencies": Map {},
  "errors": Array [
    Object {
      "message": "Cannot resolve \\"dep\\" packages
  At app
  Did you forget to run \\"esy install\\" command?",
    },
  ],
  "exportedEnv": Object {},
  "id": "app-0.1.0",
  "installCommand": Array [],
  "name": "app",
  "packageJson": Object {
    "dependencies": Object {
      "dep": "*",
    },
    "devDependencies": Object {},
    "esy": Object {
      "build": null,
      "buildsInSource": false,
      "exportedEnv": Object {},
      "install": null,
    },
    "name": "app",
    "peerDependencies": Object {},
    "version": "0.1.0",
  },
  "shouldBePersisted": false,
  "sourcePath": "",
  "sourceType": "root",
  "version": "0.1.0",
}
`;
